components/movementsensor/wheeledodometry/wheeled_odometry.go

```diff
@@ -0,0 +1,302 @@
+	RightMotors  []string `json:"right_motors"`
+	Base         string   `json:"base"`
+	TimeInterval float64  `json:"time_interval,omitempty"`
+}
+
+type motorPairs struct {
+	left  motor.Motor
+	right motor.Motor
+}
+
+type odometry struct {
+	resource.Named
+	resource.AlwaysRebuild
+
+	lastLeftPos        float64
+	lastRightPos       float64
+	baseWidth          float64
+	wheelCircumference float64
+	timeInterval       float64
+
+	motors []motorPairs
+
+	angularVelocity spatialmath.AngularVelocity
+	acceleration    r3.Vector
+	linearVelocity  r3.Vector
+	position        r3.Vector
+	orientation     spatialmath.EulerAngles
+
+	cancelFunc              func()
+	activeBackgroundWorkers sync.WaitGroup
+	mu                      sync.Mutex
+	logger                  golog.Logger
+}
+
+func init() {
+	resource.RegisterComponent(
+		movementsensor.API,
+		model,
+		resource.Registration[movementsensor.MovementSensor, *Config]{Constructor: newWheeledOdometry})
+}
+
+// Validate ensures all parts of the config are valid.
+func (cfg *Config) Validate(path string) ([]string, error) {
+	var deps []string
+
+	if cfg.Base == "" {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "base")
+	}
+	deps = append(deps, cfg.Base)
+
+	if len(cfg.LeftMotors) == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "left motors")
+	}
+	deps = append(deps, cfg.LeftMotors...)
+
+	if len(cfg.RightMotors) == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "right motors")
+	}
+	deps = append(deps, cfg.RightMotors...)
+
+	return deps, nil
+}
+
+// newWheeledOdometry returns a new wheeled encoder movement sensor defined by the given config.
+func newWheeledOdometry(
+	ctx context.Context,
+	deps resource.Dependencies,
+	conf resource.Config,
+	logger golog.Logger,
+) (movementsensor.MovementSensor, error) {
+	newConf, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+
+	timeInt := newConf.TimeInterval
+	if timeInt == 0 {
+		timeInt = defaultTimeInt
+	}
+
+	o := &odometry{
+		Named:        conf.ResourceName().AsNamed(),
+		lastLeftPos:  0,
+		lastRightPos: 0,
+		timeInterval: timeInt,
+		logger:       logger,
+	}
+
+	for i, m := range newConf.LeftMotors {
+		motorDep, err := motor.FromDependencies(deps, m)
+		if err != nil {
+			return nil, err
+		}
+		properties, err := motorDep.Properties(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+		if !properties.PositionReporting {
+			return nil, motor.NewPropertyUnsupportedError(properties, newConf.LeftMotors[i])
+		}
+		o.motors[i].left = motorDep
+	}
+
+	for i, m := range newConf.RightMotors {
+		motorDep, err := motor.FromDependencies(deps, m)
+		if err != nil {
+			return nil, err
+		}
+		properties, err := motorDep.Properties(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+		if !properties.PositionReporting {
+			return nil, motor.NewPropertyUnsupportedError(properties, newConf.RightMotors[i])
+		}
+		o.motors[i].right = motorDep
```

Comment:
Since the scope is for left motors and right motors I'm thinking I'll leave it that way for now, but maybe add warning messages if there are more than one pair of motors